Tools/peg_generator/peg_parser/pegen.c

```diff
@@ -487,90 +487,96 @@ number_token(Parser *p)
                     p->arena);
 }
 
-PyObject *
-run_parser(struct tok_state *tok, void *(start_rule_func)(Parser *), int mode, int input_mode,
-           KeywordToken **keywords, int n_keyword_lists)
+void
+Parser_Free(Parser *p)
+{
+    for (int i = 0; i < p->size; i++) {
+        PyMem_Free(p->tokens[i]);
+    }
+    PyMem_Free(p->tokens);
+    PyMem_Free(p);
+}
+
+Parser *
+Parser_New(struct tok_state *tok, mod_ty(*parse_func)(Parser *), int input_mode,
+           PyArena *arena)
 {
-    PyObject *result = NULL;
     Parser *p = PyMem_Malloc(sizeof(Parser));
     if (p == NULL) {
         PyErr_Format(PyExc_MemoryError, "Out of memory for Parser");
-        goto exit;
+        return NULL;
     }
     assert(tok != NULL);
     p->tok = tok;
     p->input_mode = input_mode;
-    p->keywords = keywords;
+    p->keywords = reserved_keywords;
     p->n_keyword_lists = n_keyword_lists;
     p->tokens = PyMem_Malloc(sizeof(Token *));
     if (!p->tokens) {
         PyErr_Format(PyExc_MemoryError, "Out of memory for tokens");
-        goto exit;
+        return NULL;
     }
     p->tokens[0] = PyMem_Malloc(sizeof(Token));
     memset(p->tokens[0], '\0', sizeof(Token));
     p->mark = 0;
     p->fill = 0;
     p->size = 1;
 
-    p->arena = PyArena_New();
-    if (!p->arena) {
-        goto exit;
+    if (arena == NULL) {
+        p->arena = PyArena_New();
+        if (!p->arena) {
+            return NULL;
+        }
+    } else {
+        p->arena = arena;
```

Comment:
I don't particularly like that you can do
```
p = Parser_New(tok, func, mode, NULL);
Parser_Free(p);
```
and now you've leaked an arena object.

Why not insist that the caller owns the arena? (Also I don't see calls that pass `NULL` so it seems unnecessary anyway?)

Reply:
Yeah this is maybe a left-over from an earlier iteration of mine.